### PR TITLE
[installer]: add support for S3-based object storage

### LINKED ISSUE #6772: [Installer]: configure and document S3 object storage
- [x] Get the installer working with S3 storage. Consider implications of proxying through Minio
- [x] Ensure works with AWS
- [ ] ~#6776~
- [x] Validate secrets
- [ ] #6704

### LINKED ISSUE #6776: [Installer]: Cannot use S3 object storage in DigitalOcean
> See [example](https://gist.github.com/MrSimonEmms/a770d6e36e3fdf8222b54626f2ae9d7d) to demonstrate the problem

DigitalOcean provides "Spaces", which is their version of S3-compatible storage. Despite having an identical API to S3, this works slightly differently to AWS S3 due to how the regions and endpoints are configured - the `endpoint` is "<regions>.digitaloceanspaces.com" - examples use `ams3`, but is the same for all regions. This is different AWS S3 which use a common `s3.amazonaws.com` and then sets the region separately.

# Attempt 1 - remove region from the URL

`endpoint := "digitaloceanspaces.com"`

This results in `XML syntax error on line 2: invalid character entity & (no semicolon)`. Likely because nothing exists on this URL

# Attempt 2 - setting region in URL and MakeBucket `Region`

The `MakeBucket` call cannot calculate the SHA256 header correctly, resulting in `The provided 'x-amz-content-sha256' header does not match what was computed`. Likely this is because DigitalOcean don't calculate the region as part of the header on their side because you cannot create, for example, an `fra1` resource on the URL `ams3`

# What works - setting region URL but not MakeBucket

If you remove the `Region` from the `MakeBucket` options, this works fine.

However, every part of Gitpod (the Installer, the ContentService etc) expects the `Region` to be set.

---

We get the same problem using Minio Gateway

# Suggested solutions

1. make `Region` an optional parameter. Theoretically, this would work, but it may have unintended consequences elsewhere especially for non-DO S3 configuration
2. don't support DigitalOcean Spaces - this would limit our self-hosted options. Furthermore, it may also be an issue with other non-AWS S3-compatible providers.

In additional to these, we could ask DigitalOcean for help and see if there's anything they can suggest.

---

# Workaround for today

Use Minio in-cluster

### PR DESCRIPTION
**Pull request by MrSimonEmms**
## Description

Add support for S3-based object storage. This also makes a change to the `content-service` to allow reading a file - this is needed as we pass in a ConfigMap, but the credentials are in a Secret.

## Related Issue(s)

Fixes #6772

## How to test

Create a secret. The username/password are an AWS IAM with `AmazonS3FullAccess` policy 

```shell
kubectl create secret generic s3-storage-token --from-literal accessKeyId=xxx --from-literal secretAccessKey=xxx
```

Update the config

```yaml
objectStorage:
  s3:
    endpoint: s3.amazonaws.com
    credentials:
      kind: secret
      name: s3-storage-token
```

Run the Installer, create a workspace and check a bucket has been created

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
@MrSimonEmms: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** after the PR has been reviewed.

Associated issue: *#6776*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by aledbf** [comment]
@MrSimonEmms do you have an ETA for this PR? I need this same feature for the EKS guide in order to support S3 storage

**Comment by MrSimonEmms** [comment]
I expect to be looking at this again tomorrow. It's part of a bigger ticket which should unblock all your requirements #6772

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** after the PR has been reviewed.

Associated issue: *#6776*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/content-service/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/content-service/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

Associated issue: *#6776*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/content-service/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/content-service/OWNERS)**
- **[components/content-service-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/content-service-api/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

Associated issue: *#6772*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/content-service/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/content-service/OWNERS)**
- **[components/content-service-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/content-service-api/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by MrSimonEmms** [comment]
When this is merged, update #6767 so the default `VERSION` in the script is that version

File: installer/pkg/common/storage.go [hunk 1]
```diff
@@ -33,13 +39,27 @@ func StorageConfig(context *RenderContext) storageconfig.StorageConfig {
 			GCloudConfig: storageconfig.GCPConfig{
 				Region:             context.Config.Metadata.Region,
 				Project:            context.Config.ObjectStorage.CloudStorage.Project,
-				CredentialsFile:    "/mnt/secrets/gcp-storage/service-account.json",
+				CredentialsFile:    getStoragePath("service-account.json"),
```
Comment (csweichel) [inline installer/pkg/common/storage.go:42]:
```suggestion
				CredentialsFile:    filepath.Join(storageMount, "service-account.json"),
```

File: installer/pkg/common/storage.go [hunk 2]
```diff
@@ -33,13 +39,27 @@ func StorageConfig(context *RenderContext) storageconfig.StorageConfig {
...
 		}
 	}
 
+	if context.Config.ObjectStorage.S3 != nil {
+		res = &storageconfig.StorageConfig{
+			Kind: storageconfig.MinIOStorage,
+			MinIOConfig: storageconfig.MinIOConfig{
+				Endpoint:            context.Config.ObjectStorage.S3.Endpoint,
+				AccessKeyIdFile:     getStoragePath("username"),
+				SecretAccessKeyFile: getStoragePath("password"),
```
Comment (csweichel) [inline installer/pkg/common/storage.go:55]:
```suggestion
				AccessKeyIdFile:     filepath.Join(storageMount, "username"),
				SecretAccessKeyFile: filepath.Join(storageMount, "password"),
```

File: installer/pkg/common/storage.go [hunk 3]
```diff
@@ -33,13 +39,27 @@ func StorageConfig(context *RenderContext) storageconfig.StorageConfig {
...
 			},
 		}
 	}
 
+	if context.Config.ObjectStorage.S3 != nil {
+		res = &storageconfig.StorageConfig{
+			Kind: storageconfig.MinIOStorage,
+			MinIOConfig: storageconfig.MinIOConfig{
+				Endpoint:            context.Config.ObjectStorage.S3.Endpoint,
+				AccessKeyIdFile:     getStoragePath("username"),
```
Comment (csweichel) [inline installer/pkg/common/storage.go:54]: Why is this file called "username" instead of `accessKeyId`?
Conversely for the `SecretAccessKey`?

**Review by csweichel — COMMENTED** [review]
a few nits

File: installer/pkg/common/storage.go (hunk 3 above)
Comment (MrSimonEmms) [reply installer/pkg/common/storage.go:54]: Minio and the S3 docs I was looking at called it user/pass and I didn't think of a better name. Will change to `accessKeyId` and `secretAccessKey`

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: f2a85809385af909745cae8ac4b98e27f8c42f67</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6777#issuecomment-978975851" title="LGTM">csweichel</a>*

Associated issue: *#6772*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/content-service/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/content-service/OWNERS)~~ [csweichel]
- ~~[components/content-service-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/content-service-api/OWNERS)~~ [csweichel]
- ~~[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**